src/Illuminate/Log/Context/Repository.php

```diff
@@ -0,0 +1,421 @@
+                ...[$key => $this->getSecret($key)],
+            ];
+        }, []);
+    }
+
+    /**
+     * Push the given values onto the key's stack.
+     *
+     * @param  string  $key
+     * @param  mixed  ...$values
+     * @return $this
+     */
+    public function push($key, ...$values)
+    {
+        if (! $this->isStackable($key)) {
+            throw new RuntimeException("Unable to push value onto context stack for key [{$key}].");
+        }
+
+        $this->data[$key] = [
+            ...$this->data[$key] ?? [],
+            ...$values,
+        ];
+
+        return $this;
+    }
+
+    /**
+     * Push the given values onto the key's secret stack.
+     *
+     * @param  string  $key
+     * @param  mixed  ...$values
+     * @return $this
+     */
+    public function pushSecret($key, ...$values)
+    {
+        if (! $this->isSecretStackable($key)) {
+            throw new RuntimeException("Unable to push value onto secret context stack for key [{$key}].");
+        }
+
+        $this->secrets[$key] = [
+            ...$this->secrets[$key] ?? [],
+            ...$values,
+        ];
+
+        return $this;
+    }
+
+    /**
+     * Determine if the given key exists.
+     *
+     * @param  string  $key
+     * @return bool
+     */
+    public function has($key)
+    {
+        return array_key_exists($key, $this->data);
+    }
+
+    /**
+     * Determine if the given key exists as a secret.
+     *
+     * @param  string  $key
+     * @return bool
+     */
+    public function hasSecret($key)
+    {
+        return array_key_exists($key, $this->secrets);
+    }
+
+    /**
+     * Remember the authenticated user.
+     *
+     * @return $this
+     */
+    public function rememberAuthUser()
+    {
+        $this->app['events']->listen([Login::class, Authenticated::class], fn (Login|Authenticated $event) => $this->addSecret('auth', [
+            'id' => $event->user->getAuthIdentifier(),
+            'guard' => $event->guard,
+        ]));
+
+        $this->app['events']->listen(fn (Logout $event) => $this->forgetSecret('auth'));
+
+        $this->app['events']->listen(function (Hydrated $event) {
+            $auth = $this->app['auth']->forgetGuards();
+
+            $this->restoreGuard ??= $auth->getDefaultDriver();
+
+            if (! $this->hasSecret('auth')) {
+                $auth->shouldUse($this->restoreGuard);
+            }
+
+            ['id' => $id, 'guard' => $guard] = $this->getSecret('auth');
+
+            if (method_exists($auth->guard($guard), 'getProvider')) {
+                $provider = $auth->guard($guard)->getProvider();
+            } elseif ($providerName = $this->app['config']->has("auth.guards.{$guard}.provider")) {
+                $provider = $auth->createUserProvider($providerName);
+            } else {
+                $auth->shouldUse($this->restoreGuard);
+
+                return;
+            }
+
+            tap($auth)->shouldUse($guard)->setUser($provider->retrieveById($id));
```

Comment:
One thing I don't love here is that the user is eagerly retrieved.

I've been wondering if we should introduce a "ContextGuard" that would only resolve the user when it is being accessed.

That way if the `Auth::user` method is invoked we do not perform a database query.

Thoughts?

Reply:
Having this would be great - atleast in applications that are performance critical. So, all context does not become available by default - or rather one could say that this Context value that is being set should only be resolved if the value is being accessed.

Reply:
One distinction I make on my Laravel projects is the separation of User Eloquent and the Authenticatable object. That way, once I build an Authenticatable object, it's a plain PHP Class that can be serialized / deserialized and can be passed through execution context without causing database load.

This allows (for instance) situations where I pass my Authenticatable object into a Job class and the Laravel Worker doesn't even establishes a database connection since the job itself is more concerned with Elasticsearch, S3 or unrelated database operations. It's similar to the concept of JWT claims having all the authenticated information statelessly.

Reply:
I've introduced a `Context` guard to all the user to be lazily resolved. Now there is no DB interactions until needed.

@deleugpn, your approach is certainly a great approach, but I don't think that is super common for Laravel apps. I think most Laravel applications would be expecting a model from `Auth::user()`.

Context still supports your usecase:

```php
Context::dehydrating(function () {
    if ($user = Auth::user()) {
        Context::addSecret('auth', serialize($user));
    }
});

Context::hydrated(function () {
    if ($user = Context::getSecret('auth')) {
        Auth::setUser(unserialize($user));
    }
});

// If you are using a stateful guard...

Event::listen(function (Logout $event) {
    Context::forgetSecret('auth');
});
```